### ISSUE TITLE
`attr-defined` false positive for module attribute imported from another module

### ISSUE THREAD

**Issue by markshannon** [python/mypy#16596]
I'd be surprised if I'm the first person to report this, but I can't find an issue for it.

**Bug Report**

False positive for `attr-defined`: Module "interface" does not explicitly export attribute "foo"

"explicitly export" appears to mean "appears in `__all__`"

**To Reproduce**

`internal.py`:
```Py
def foo(): pass
```
`interface.py`:
```Py
from internal import foo
```
and:
```Py
import interface
interface.foo # attr-defined error here: Module "interface" does not explicitly export attribute "foo"
```

**Expected Behavior**

No error

Whether the above code is good code is debatable, and maybe a linter should flag this.
However, MyPy is a type checker and the above code is type safe. There should be no error.

**Actual Behavior**

Module "interface" does not explicitly export attribute "foo"

**Comment by aganders3** [comment]
> "explicitly export" appears to mean "appears in __all__"

You can also use `from internal import foo as foo` to "explicitly" export without using `__all__`.

**Comment by AlexWaygood** [comment]
(Marked as pending, since this is an optional check that is disabled by default, and can still easily be disabled even if you have the rest of `strict = True` enabled.)

**Comment by AlexWaygood** [comment]
> Should I open a new issue for having better options for --strict or is #16599 sufficient?

I think #16599 is sufficient, though I think specific changes will have to be considered in isolation. For example: I actually agree it would be a good user experience to have `--allow-redefinition` enabled by default (though in my opinion, it should be disabled if the user specifies `--strict`). But I can't see how we could remove `--disallow-implicit-reexport` from the options specified by `--strict` while keeping backwards compatibility for existing users of `--strict`.

I don't have the final call here; I'm just a triager, not a mypy maintainer. I doubt you'll be able to find a maintainer willing to reopen this, however :)

**Comment by pktiuk** [comment]
I have this issue and setting `implicit_reexport = true` does not help with it.

```py
from django.contrib.gis import admin

class CustomAdmin(admin.GISModelAdmin):
    @admin.display(description="Some description")
    def something(self, instance: MyModel) -> str:
        pass #......
```

`django.contrib.gis.admin` contains:

```py
from django.contrib.admin import (
    HORIZONTAL,
    VERTICAL,
    AdminSite,
    ModelAdmin,
    StackedInline,
    TabularInline,
    action,
    autodiscover,
    display,
    register,
    site,
)
from django.contrib.gis.admin.options import GISModelAdmin

__all__ = [
    "HORIZONTAL",
    "VERTICAL",
    "AdminSite",
    "ModelAdmin",
    "StackedInline",
    "TabularInline",
    "action",
    "autodiscover",
    "display",
    "register",
    "site",
    "GISModelAdmin",
]
```


Does it match the scope of this issue, or maybe it should be considered as an another bug?
